src/SQLite.cs

```diff
@@ -3158,6 +3167,229 @@ object ReadCol (Sqlite3Statement stmt, int index, SQLite3.ColType type, Type clr
+				});
+			}
+			else if (clrType == typeof (sbyte)) {
+				fastSetter = CreateNullableTypedSetterDelegate<T, sbyte> (column, (stmt, index) => {
+					return (sbyte)SQLite3.ColumnInt (stmt, index);
+				});
+			}
+			else if (clrType == typeof (byte[])) {
+				fastSetter = CreateTypedSetterDelegate<T, byte[]> (column, (stmt, index) => {
+					return SQLite3.ColumnByteArray (stmt, index);
+				});
+			}
+			else if (clrType == typeof (Guid)) {
+				fastSetter = CreateNullableTypedSetterDelegate<T, Guid> (column, (stmt, index) => {
+					var text = SQLite3.ColumnString (stmt, index);
+					return new Guid (text);
+				});
+			}
+			else if (clrType == typeof (Uri)) {
+				fastSetter = CreateTypedSetterDelegate<T, Uri> (column, (stmt, index) => {
+					var text = SQLite3.ColumnString (stmt, index);
+					return new Uri (text);
+				});
+			}
+			else if (clrType == typeof (StringBuilder)) {
+				fastSetter = CreateTypedSetterDelegate<T, StringBuilder> (column, (stmt, index) => {
+					var text = SQLite3.ColumnString (stmt, index);
+					return new StringBuilder (text);
+				});
+			}
+			else if (clrType == typeof (UriBuilder)) {
+				fastSetter = CreateTypedSetterDelegate<T, UriBuilder> (column, (stmt, index) => {
+					var text = SQLite3.ColumnString (stmt, index);
+					return new UriBuilder (text);
+				});
+			}
+			else {
+				// NOTE: Will fall back to the slow setter method in the event that we are unable to create a fast setter delegate for a particular column type
+			}
+			return fastSetter;
+		}
+
+		/// <summary>
+		/// This creates a strongly typed delegate that will permit fast setting of column values given a Sqlite3Statement and a column index.
+		///
+		/// Note that this is identical to CreateTypedSetterDelegate(), but has an extra check to see if it should create a nullable version of the delegate.
+		/// </summary>
+		/// <typeparam name="ObjectType">The type of the object whose member column is being set</typeparam>
+		/// <typeparam name="ColumnMemberType">The CLR type of the member in the object which corresponds to the given SQLite columnn</typeparam>
+		/// <param name="column">The column mapping that identifies the target member of the destination object</param>
+		/// <param name="getColumnValue">A lambda that can be used to retrieve the column value at query-time</param>
+		/// <returns>A strongly-typed delegate</returns>
+		private static Action<ObjectType, Sqlite3Statement, int> CreateNullableTypedSetterDelegate<ObjectType, ColumnMemberType> (TableMapping.Column column, Func<Sqlite3Statement, int, ColumnMemberType> getColumnValue) where ColumnMemberType : struct
+		{
+			var clrTypeInfo = column.PropertyInfo.PropertyType.GetTypeInfo();
+			bool isNullable = false;
+
+			if (clrTypeInfo.IsGenericType && clrTypeInfo.GetGenericTypeDefinition () == typeof (Nullable<>)) {
+				isNullable = true;
+			}
+
+			if (isNullable) {
+				var setProperty = (Action<ObjectType, ColumnMemberType?>)Delegate.CreateDelegate (
```

Comment:
It's frustrating to me how much of this code is almost identically copy-pasted between `CreateNullableTypedSetterDelegate` and `CreateTypedSetterDelegate`, but I couldn't find a clean way around it.  I don't know if / how ProtoBuf deals with this, but I wonder if they don't support nullable types the way SQLite-net does.

Reply:
Yeah. I like your code, my biggest hesitation is how much dupe code the reader layers are accumulating. I think that cleanup work will be for another time though.